netket/optimizer/sr/s_jacobian_mat.py

```diff
@@ -0,0 +1,205 @@
+    for all samples σ_i at given values of the parameters p_j
+    Average <O_j> subtracted for each parameter
+    Divided through by sqrt(#samples) to normalise S matrix
+    If scale is not None, normalised to unit magnitude"""
+
+    scale: Optional[jnp.ndarray] = None
+    """If not None, gives scale factors with which O is normalised"""
+
+    x0: Optional[PyTree] = None
+    """Optional initial guess for the iterative solution."""
+
+    @jax.jit
+    def __matmul__(self, vec: Union[PyTree, jnp.ndarray]) -> Union[PyTree, jnp.ndarray]:
+        if not hasattr(vec, "ndim"):
+            vec, unravel = nkjax.tree_ravel(vec)
+        else:
+            unravel = lambda x: x
+
+        return unravel(
+            jnp.transpose(jnp.conj(jnp.transpose(jnp.conj(S.O @ vec)) @ S.O))
+            + S.sr.diag_shift * vec
+        )
+
+    @jax.jit
+    def solve(self, y: PyTree, x0: Optional[PyTree] = None) -> PyTree:
+        """
+        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.
+
+        Args:
+            y: the vector y in the system above.
+            x0: optional initial guess for the solution.
+
+        Returns:
+            x: the PyTree solving the system.
+            info: optional additional informations provided by the solver. Might be
+                None if there are no additional informations provided.
+        """
+
+        # Ravel input PyTrees, record unravelling function too
+        grad, unravel = nkjax.tree_ravel(grad)
+        
+        if x0 is None:
+            x0 = self.x0
+        if x0 is not None:
+            x0, _ = nkjax.tree_ravel(x0)
+            if self.scale is not None:
+                x0 = x0 / self.scale
+
+        if self.scale is not None:
+            grad = grad * self.scale
+
+        solve_fun = self.sr.solve_fun()
+        _mat_vec = lambda x: self @ x
+        out, info = solve_fun(_mat_vec, grad, x0=x0)
+
+        if self.scale is not None:
+            out = out * self.scale
+
+        return unravel(out), info
+
+    @jax.jit
+    def to_dense(self) -> jnp.ndarray:
+        """
+        Convert the lazy matrix representation to a dense matrix representation.
+
+        Returns:
+            A dense matrix representation of this S matrix.
+        """
+        return jnp.transpose(jnp.conj(self.O)) @ self.O + self.sr.diag_shift * jnp.eye(
+            self.O.shape[1]
+        )
+
+
+
+@partial(jax.jit, static_argnums=(0, 4, 5))
+def gradients(
+    apply_fun: Callable[[PyTree, jnp.ndarray], jnp.ndarray],
+    params: PyTree,
+    samples: jnp.ndarray,
+    model_state: Optional[PyTree],
+    mode: str,
+    rescale_shift: bool,
+):
+    """Calculates the gradients O_ij by backpropagating every sample separately,
+    vectorising the loop using vmap
+    If rescale_shift is True, columns of O are rescaled to unit magnitude, and
+    scale factor [1/sqrt(S_kk)] returned as a separate vector for
+    scale-invariant regularisation as per Becca & Sorella p. 143."""
```

Comment:
```suggestion
    scale-invariant regularisation as per Becca & Sorella p. 143.
    """
```

And could you also give more details about what `mode` does? what changes between R2R, R2C and HOLO?

Reply:
I've expanded the docstring of `SRJacobian.mode`, please check if that's what you had in mind